osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/security/ReplaceKeyCommandExecutor.java

```diff
@@ -66,146 +49,29 @@ public ReplaceKeyCommandExecutor() {
-  }
-
   @Override
-  public ActionResponseDto executeBundleAction(
+  public ActionResponseDto execute(
       final DlmsConnectionManager conn,
       final DlmsDevice device,
-      final ActionRequestDto actionRequestDto,
+      final SetKeysRequestDto setKeysRequestDto,
       final MessageMetadata messageMetadata)
       throws OsgpException {
+    this.checkActionRequestType(setKeysRequestDto);
 
-    this.checkActionRequestType(actionRequestDto);
-
-    LOGGER.info("Keys set on device :{}", device.getDeviceIdentification());
-
-    SetKeysRequestDto setKeysRequestDto = (SetKeysRequestDto) actionRequestDto;
-
-    if (!setKeysRequestDto.isGeneratedKeys()) {
-      setKeysRequestDto = this.decryptRsaKeys((SetKeysRequestDto) actionRequestDto);
-    }
-    // if keys are generated, then they are unencrypted by the GenerateAndReplaceKeyCommandExecutor
-
-    final DlmsDevice devicePostSave =
-        this.execute(
-            conn,
-            device,
-            ReplaceKeyCommandExecutor.wrap(
-                setKeysRequestDto.getAuthenticationKey(),
-                KeyId.AUTHENTICATION_KEY,
-                SecurityKeyType.E_METER_AUTHENTICATION,
-                setKeysRequestDto.isGeneratedKeys()),
-            messageMetadata);
-
-    this.execute(
-        conn,
-        devicePostSave,
-        ReplaceKeyCommandExecutor.wrap(
-            setKeysRequestDto.getEncryptionKey(),
-            KeyId.GLOBAL_UNICAST_ENCRYPTION_KEY,
-            SecurityKeyType.E_METER_ENCRYPTION,
-            setKeysRequestDto.isGeneratedKeys()),
-        messageMetadata);
-
-    return new ActionResponseDto(REPLACE_KEYS + device.getDeviceIdentification() + WAS_SUCCESFULL);
-  }
-
-  private SetKeysRequestDto decryptRsaKeys(final SetKeysRequestDto setKeysRequestDto)
-      throws FunctionalException {
     try {
-      final byte[] authenticationKey =
-          this.decrypterForGxfSmartMetering.decrypt(setKeysRequestDto.getAuthenticationKey());
-      final byte[] encryptionKey =
-          this.decrypterForGxfSmartMetering.decrypt(setKeysRequestDto.getEncryptionKey());
-
-      return new SetKeysRequestDto(authenticationKey, encryptionKey);
-    } catch (final EncrypterException e) {
-      throw new FunctionalException(
-          FunctionalExceptionType.DECRYPTION_EXCEPTION, ComponentType.PROTOCOL_DLMS, e);
-    }
-  }
-
-  @Override
-  public DlmsDevice execute(
-      final DlmsConnectionManager conn,
-      final DlmsDevice device,
-      final ReplaceKeyInput keyWrapper,
-      final MessageMetadata messageMetadata)
-      throws OsgpException {
-
-    if (!keyWrapper.isGenerated()) {
-      this.secretManagementService.storeNewKey(
-          messageMetadata,
-          device.getDeviceIdentification(),
-          keyWrapper.getSecurityKeyType(),
-          keyWrapper.getBytes());
+      this.deviceKeyProcessingService.startProcessing(messageMetadata.getDeviceIdentification());
```

Comment:
I see this execute method with the startProcessing call also in the abstract super class. Either implement it in the super and use that or remove the implementation in the abstract super and use the specific implementation.